proof-language.md

```diff
@@ -0,0 +1,359 @@
+    right: Pattern
+
+class Exists(Pattern)
+    var: EVar
+    subpattern : Pattern
+
+class Mu(Pattern)
+    var: SVar
+    subpattern : Pattern
+
+    def well_formed():
+        return super.well_formed()
+           and subpattern.var_occurs_positively(var)
+```
+
+### Meta-patterns
+
+Meta-patterns allow us to represent axiom- and theorem-schemas through the use
+of metavariables:
+
+```python
+class MetaVar : Pattern
+    name: uint32
+
+    # Meta-requirements that must be satisfied by any instantiation.
+    fresh: list[EVar | SVar]        # variables that must not occur in an instatiation
+    positive: list[SVar]            # variables that must only occur positively in an instatiation
+    negative: list[SVar]            # variables that must only occur negatively in an instatiation
+    application_context: list[EVar] # Variables that must only occur as a hole variable in an application context.
+```
+
+Each `MetaVar` has a list of requirements that must be met by any instantiation.
+These may also be used by the well-formedness checks for `Proof`s.
+
+We also need to represent substitutions applied to `MetaVar`s.
+
+```python
+class ESubst : Pattern
+    pattern: MetaVar
+    var: EVar
+    plug: Pattern
+
+class SSubst : Pattern 
+    pattern: MetaVar
+    var: SVar
+    plug: Pattern
+```
+
+
+`Proof`s
+--------
+
+### Axiom Schemas
+
+Axiom schemas are `Proof`s that do not need any input arguments.
+They may use `MetaVar`s to represent their schematic nature.
+
+```python
+class Lukaseiwicz : Proof
+    def conclusion():
+        phi1 = MetaVar('phi1')
+        return Implication(Implication(Implication(MetaVar(phi1) , ...)...)...)
+
+class Quantifier : Proof
+    def conclusion():
+        x = EVar('#x')
+        y = EVar('#y')
+        phi = MetaVar('phi', fresh=[y])
+        return Implication(ESubst(phi, x, y), Exists(x, phi))
+
+class PropagationOr : Proof
+    def conclusion():
+        hole = EVar('#hole')
+        C = MetaVar(application_context=(EVar('#hole'),))
+        phi1 = MetaVar('#phi1')
+        phi2 = MetaVar('#phi2')
+        return Implication(ESubst(C, or(phi1, phi2), hole), or(ESubst(C, phi1, hole), ESubst(C, phi2, hole)))
+
+...
+```
+
+
+### Meta Inference rule
+
+Using a single rule of meta inference, we may instantiate metavariables.
+Note that we do not need to instantiate metavariables immediately.
+This allows us to prove theorem schemas, such as $\phi \limplies \phi$.
+
+```python
+class InstantiateSchema : Proof
+    subproof : Proof
+    metavar: MetaVar
+    instantiation: Pattern
+
+    def well_formed():
+        # Fails if the instantiation does not meet the
+        # disjoint/positive/freshness/application ctx
+        # criterea of the metavar.
+
+    def conclusion():
+        return subproof.meta_substitute(metavar, instantiation)
+```
+
+### Ordinary inference
+
+```python
+class ModusPonens: Proof
+    premise_left: Implication
+    premise_right: Pattern
+
+    def conclusion():
```

Comment:
MP should also have a well-formedness check that checks well-formedmness of the premises. This should also be in Generalization, for example.

It seems to me that also, this check should actually be called when calling `conclusion()`. This is for the inductive argument that a proof is well-formed.